test/unit/config/test_path_resolves_to.py

```diff
@@ -0,0 +1,141 @@
+import pytest
+
+from galaxy.config import GalaxyAppConfiguration
+
+
+MOCK_DEPRECATED_DIRS = {
+    'my_config_dir': 'old-config',
+    'my_data_dir': 'old-database',
+}
+
+# Mock properties loaded from schema (all options represent paths)
+MOCK_SCHEMA = {
```

Comment:
You could have just loaded the actual Galaxy schema and it would have been a better test IMO - this much mocking is increasing the test isolation but making the test more brittle, verbose and as a result harder to maintain. I'd encourage you to mock less and broaden what you think about as a unit - mocking out expensive interactions and not every interaction. I've read many articles, books that argue both ways and experimented with both kinds of tests in this code base and I've just developed an intuition that this is too much mocking. I think the most valuable tests in "test/unit" either look more like quasi-integration tests or are testing really complex "units".

Given that this is just an opinion and likely people on the team even would disagree with me - maybe I'd just encourage writing a bit more of integration test with your next feature and see if it feels like it would catch more errors/be easier to maintain - and do this mock heavy testing again if it doesn't feel right. This testing is certainly way better than no testing.

For this code base and functionality like this - I'd rather see just one integration test where a couple paths are correctly rewritten - and then break down into unit tests like this when there are more than a couple edge cases, complex error handling, etc..

Reply:
I trust your intuition, as my experience with large code bases is limited. And I do think this should be covered with integration tests as well.

I try to restrain myself from doing clever things in test code. I'll be more conservative going forward.

I still feel that in this particular case (and similar cases) we need simple tests that isolate specific functionality. So, my motivation is not so much to keep IO out, but rather to be able to set a very specific context (which is not always present in the real app) - and test that that context is handled as expected.

Maintainability. These tests use generic property names (I used my_config_dir instead of config_dir on purpose), so they test not specific properties, but the interaction of properties - which is exactly what I want to test. Also, using this staged context prevents the test from breaking when a concrete property name is changed.

All that said, I do recognize and agree with all the downsides you mention. Going forward, I'll keep this in mind and will try to evaluate carefully the tradeoffs between test isolation and the complexity mocking adds w.r.t. maintaining tests (and ensuring they continue testing what they are expected to test over time). Also will do less verbose, limited to what's critical, etc. And I'll try to add integration tests where appropriate.